internal/pkg/gateway/registry/registry.go

```diff
@@ -7,40 +7,307 @@ SPDX-License-Identifier: Apache-2.0
+	if err != nil {
+		return nil, err
+	}
+
+	for _, eps := range config.GetOrderers() {
+		for _, ep := range eps.Endpoint {
+			url := fmt.Sprintf("%s:%d", ep.Host, ep.Port)
+			if ord, ok := reg.orderers[url]; ok {
+				orderers = append(orderers, ord)
+			}
+		}
+	}
+
+	return orderers, nil
+}
+
+func (reg *Registry) addPeer(mspid string, url string) error {
+	// assumes that the msp registry has already been populated with this mspid
+	if _, ok := reg.peers[url]; !ok && strings.Contains(url, ":") {
+		host, _, err := net.SplitHostPort(url)
+		if err != nil {
+			return err
+		}
+		// this peer is new - connect to it and add to the peers registry
+		endorser, err := reg.newRemoteEndorser(url, host, mspid)
+		if err != nil {
+			return err
+		}
+		reg.peers[url] = endorser
+		reg.logger.Infof("Added peer to registry: %s", url)
+	}
+
+	return nil
+}
+
+func (reg *Registry) addOrderer(mspid string, host string, port uint32) error {
+	// assumes that the msp registry has already been populated with this mspid
+	url := fmt.Sprintf("%s:%d", host, port)
+	if _, ok := reg.orderers[url]; !ok {
+		// this orderer is new - connect to it and add to the orderers registry
+		orderer, err := reg.newOrderer(url, host, mspid)
+		if err != nil {
+			return err
+		}
+		reg.orderers[url] = orderer
+		reg.logger.Infof("Added orderer to registry: %s", url)
+	}
+	return nil
+}
+
+func (reg *Registry) registerChannel(channel string) error {
+	// todo need to handle membership updates
+	init, ok := reg.channelsInitialized[channel]
+	if !init || !ok {
+		// get config and peer discovery info for this channel
+
+		// Get the config
+		config, err := reg.discovery.Config(channel)
+		if err != nil {
+			return err
+		}
+		// get the tlscerts
+		for msp, info := range config.GetMsps() {
+			reg.msps[msp] = info.GetTlsRootCerts()[0] // todo handle more than one
+		}
+
+		for msp, eps := range config.GetOrderers() {
+			for _, ep := range eps.Endpoint {
+				reg.addOrderer(msp, ep.Host, ep.Port)
+			}
+		}
+
+		// get the peers for the channel
+		peers := map[string]string{}
+		members := reg.discovery.PeersOfChannel(common.ChannelID(channel))
+		for _, member := range members {
+			id := member.PKIid.String()
+			peers[id] = member.PreferredEndpoint()
+		}
+		for org, infoset := range reg.discovery.IdentityInfo().ByOrg() {
+			for _, info := range infoset {
+				pkid := info.PKIId.String()
+				if endpoint, ok := peers[pkid]; ok {
+					reg.addPeer(org, endpoint)
+				}
+			}
+		}
+		reg.channelsInitialized[channel] = true
+	}
+	return nil
+}
+
+func (reg *Registry) newTLSConnection(url, host, mspid string) (*grpc.ClientConn, error) {
+	tlscert, ok := reg.msps[mspid]
+	if !ok {
+		return nil, fmt.Errorf("unknown msp: %s", mspid)
+	}
+	certPool := x509.NewCertPool()
+	if !certPool.AppendCertsFromPEM(tlscert) {
+		return nil, fmt.Errorf("failed to append certificate to client credentials")
+	}
+	creds := credentials.NewClientTLSFromCert(certPool, host)
```

Comment:
1. It's better to use the [already existing](https://github.com/hyperledger/fabric/blob/master/internal/pkg/comm/client.go) client to create connections.
2. Perhaps consider decoupling the communication from the registry? All you need is some kind of "factory" for NewAtomicBroadcastClient right?

Reply:
Thanks for the suggestions, Yacov.  I've made both these changes.